File: sdk/python/provenance_client/interfaces/session.py
from typing import Any, Callable, Protocol

from ..core.models import Decision, ExecutionResult


class ProvenanceSessionProtocol(Protocol):
    def guard(
        self,
        action: str | None = None,
        **kwargs,
    ) -> Callable[[Any], Any]: ...

    def execute(
        self,
        action: str,
        parameters: dict,
        *,
        decision: Decision | None = None,
    ) -> ExecutionResult: ...

    async def async_execute(
        self,
        action: str,
        parameters: dict,
        *,
        decision: Decision | None = None,
    ) -> ExecutionResult: ...

    def __enter__(self) -> "ProvenanceSessionProtocol": ...

    def __exit__(self, exc_type, exc_val, exc_tb) -> None: ...

File: sdk/typescript/provenance_client/interfaces/session.ts
import { Decision, ExecutionResult } from "../core/models";

export interface ProvenanceSessionProtocol {
  guard(
    action?: string,
    options?: {
      decision?: Decision;
      raiseOnBlock?: boolean;
    },
  ): <T extends (...args: any[]) => any>(func: T) => T;

  execute(
    action: string,
    parameters: Record<string, any>,
    options?: {
      decision?: Decision;
    },
  ): ExecutionResult;

  asyncExecute(
    action: string,
    parameters: Record<string, any>,
    options?: {
      decision?: Decision;
    },
  ): Promise<ExecutionResult>;

  readonly results: ExecutionResult[];
  readonly blockedCount: number;
  readonly allowedCount: number;

  [Symbol.dispose](): void;
}
